Tundra doesn't respect 260 path limit on windows

If you use relative pathes in tundra.lua it's easy to end up in situation when absolute path to obj file is longer than 260 characters. In this case tundra doesn't complain until linker stage, when linking fails because linker cannot access that object file.

I can't find obvious workaround for this ... it's probably override function that generate path to obj files.

I failed to add any hash lib (all of them require bit/bit32 module, which is not available under tundra lua). So I'm doing this :

    local relative_name = drop_suffix(object_fn:gsub("%.%.", "dotdot"))
    path, relative_name2 = relative_name:match("(.-)([^\\/]-%.?([^%.\\/]*))$")
    if relative_name2:len() > 0 then
      relative_name = "__" .. relative_name2
    end

And getting this everytime :

    LINK : fatal error LNK1181: cannot open input file '.obj'

Any ideas why is that ?

So I've finally got time to fix it, my solution is follows :

	diff --git a/path.lua b/path.lua
	index f5d7068..7381a35 100644
	--- a/path.lua
	+++ b/path.lua
	@@ -1,6 +1,7 @@
	module(..., package.seeall)
	
	local npath = require "tundra.native.path"
	+local native = require "tundra.native"
	
	split             = npath.split
	normalize         = npath.normalize
	@@ -20,6 +21,14 @@ function remove_prefix(prefix, fn)
	end
	end
	
	+function strip_unknown_symbols(str)
	+  local s = ""
	+  for i in str:gmatch( "%C+" ) do
	+    s = s .. i
	+  end
	+  return s
	+end
	+
	function make_object_filename(env, src_fn, suffix)
	local object_fn
	
	@@ -43,8 +52,9 @@ function make_object_filename(env, src_fn, suffix)
	-- the DAG)
	do
		local relative_name = drop_suffix(object_fn:gsub("%.%.", "dotdot"))
	+    relative_name = strip_unknown_symbols(native.digest_guid(relative_name))
		object_fn = "$(OBJECTDIR)/$(UNIT_PREFIX)/" .. relative_name .. "__" .. src_suffix .. suffix
	end
	
	return object_fn
	-end
	+end
	\ No newline at end of file

I had hard time previously because I haven't noticed that digest_guid add "\0" to string, so all appended strings "failed", because of this I had this error for a long time :

	LINK : fatal error LNK1181: cannot open input file '.obj'

This solution is required if you have folder structure similar to :

	C:/folder_a/name/name/.../name/name/test_1.cpp
	C:/folder_a/name/name/.../name/name/test_2.cpp
	...
	C:/folder_a/name/name/.../name/name/test_N.cpp
	C:/folder_b/name/name/.../name/name/tundra.lua

Because tundra.lua root folder is different (folder_b vs folder_a), tundra will create obj files with relative path, and this path will be longer then 260 characters.

Not sure if this solution is needed for upstream, but could be useful :)

--- a/scripts/tundra/path.lua
+++ b/scripts/tundra/path.lua
@@ -1,6 +1,7 @@
 module(..., package.seeall)
 
 local npath = require "tundra.native.path"
+local native = require "tundra.native"
 
 split             = npath.split
 normalize         = npath.normalize
@@ -11,6 +12,7 @@ get_extension     = npath.get_extension
 drop_suffix       = npath.drop_suffix
 get_filename_base = npath.get_filename_base
 is_absolute       = npath.is_absolute
+digest_guid       = native.digest_guid
 
 function remove_prefix(prefix, fn)
   if fn:find(prefix, 1, true) == 1 then
@@ -42,7 +44,7 @@ function make_object_filename(env, src_fn, suffix)
   -- up clobbering each other (Tundra emits an error for this when checking
   -- the DAG)
   do
-    local relative_name = drop_suffix(object_fn:gsub("%.%.", "dotdot"))
+    local relative_name = digest_guid(drop_suffix(object_fn:gsub("%.%.", "dotdot")))
     object_fn = "$(OBJECTDIR)/$(UNIT_PREFIX)/" .. relative_name .. "__" .. src_suffix .. suffix
   end
 

